index.js

```diff
@@ -219,88 +204,92 @@ async function handlePaidScheduledRewards (req, res, redis, signerAddresses, red
+      ) AS bulk
+      WHERE scheduled_rewards.address = bulk.address
+    `, [body.participants, scheduledRewardsDeltas])
+    await pgClient.query(`
+      INSERT INTO logs (address, scheduled_rewards_delta)
+      VALUES (UNNEST($1::TEXT[]), UNNEST($2::NUMERIC[]))
+    `, [body.participants, scheduledRewardsDeltas])
+    await pgClient.query(`
+      DELETE FROM logs WHERE timestamp < NOW() - INTERVAL '30 days'
+    `)
+    await pgClient.query('COMMIT')
+  } catch (err) {
+    await pgClient.query('ROLLBACK')
+    if (err.constraint === 'amount_not_negative') {
+      httpAssert.fail(400, `Scheduled rewards would become negative: ${err.detail}`)
     }
-    await tx.exec()
+    throw err
   } finally {
-    await lock.unlock()
+    pgClient.release()
   }
 
+  logger.info(`Increased scheduled rewards of ${body.participants.length} participants`)
+
   logger.info(`Merked scheduled rewards of ${body.participants.length} participants as paid`)
+
+  const { rows } = await pgPool.query(`
+    SELECT address, amount
+    FROM scheduled_rewards
+    WHERE address = ANY($1)  
+  `, [body.participants])
   json(
     res,
-    Object.fromEntries(
-      body.participants.map((address, i) => [
-        address,
-        String(updatedRewards[i])
-      ])
-    )
+    Object.fromEntries(rows.map(({ address, amount }) => [address, String(amount)]))
   )
 }
 
-async function handleGetAllScheduledRewards (res, redis) {
-  json(res, await redis.hgetall('rewards'))
+async function handleGetAllScheduledRewards (res, pgPool) {
+  const { rows } = await pgPool.query(`
+    SELECT address, amount
+    FROM scheduled_rewards
+  `)
+  json(
+    res,
+    Object.fromEntries(rows.map(({ address, amount }) => [address, String(amount)]))
+  )
 }
 
-async function handleGetSingleScheduledRewards (req, res, redis) {
+async function handleGetSingleScheduledRewards (req, res, pgPool) {
   const address = req.url.split('/').pop()
+  const { rows } = await pgPool.query(`
+    SELECT amount
+    FROM scheduled_rewards
+    WHERE address = $1
+  `, [address])
   json(
     res,
-    (await redis.hget('rewards', address)) || '0'
+    String(rows[0]?.amount || '0')
   )
 }
 
-async function handleGetLog (res, redis) {
+async function handleGetLog (res, pgPool) {
   res.setHeader('Content-Type', 'application/json')
-  res.write('[')
-
-  // Fetch logs in batches to avoid upstash request size limit
-  // https://upstash.com/docs/redis/troubleshooting/max_request_size_exceeded
-  let offset = 0
-  const batchSize = 1000
-  while (true) {
-    const batch = await redis.lrange('log', offset, offset + batchSize - 1)
-    if (batch.length === 0) {
-      break
-    }
-    if (offset !== 0) {
-      res.write(',')
-    }
-    res.write(batch.join(','))
-    offset += batchSize
-  }
-
-  res.end(']')
+  const { rows } = await pgPool.query(`
+    SELECT timestamp, address, score, scheduled_rewards_delta
+    FROM logs
+  `)
+  json(res, rows.map(row => ({
```

Comment:
Are you sure we can load all logs into the memory?

See the comment you removed:

```js
// Fetch logs in batches to avoid upstash request size limit
// https://upstash.com/docs/redis/troubleshooting/max_request_size_exceeded
```

We may need to stream the response, see https://node-postgres.com/apis/cursor

> A cursor can be used to efficiently read through large result sets without loading the entire result-set into memory ahead of time. It's useful to simulate a 'streaming' style read of data, or exit early from a large result set.

Reply:
But you're right that this needs to be looked at

Reply:
7ba7d41